cleverhans/utils.py

```diff
@@ -220,6 +220,101 @@ def grid_visual(data):
     return figure
 
 
+def get_logits_over_interval(sess, model, x, adv_x, x_data,
+                             min_epsilon=-10, max_epsilon=10,
+                             num_points=21):
+    """Get logits when the input is perturbed in an interval in adv direction.
+
+    Args:
+        sess: Tf session
+        model: Model for which we wish to get logits
+        x: Data tensor of shape [None, height, width, channels]
+        adv_x: Tensor for adversarial perturbation of x
+        x_data: Numpy array corresponding to single data
+                point of shape [height, width, channels].
+        min_epsilon: Minimum value of epsilon over the interval
+        max_epsilon: Maximum value of epsilon over the interval
+        num_points: Number of points used to interpolate
+
+    Returns:
+        Numpy array containing logits
+
+    Raises:
+        ValueError if min_epsilon is larger than max_epsilon
+    """
+    import tensorflow as tf
+    if min_epsilon > max_epsilon:
+        raise ValueError('Minimum epsilon is less than maximum epsilon')
+
+    eta = tf.nn.l2_normalize(adv_x - x, dim=0)
+    x_data = np.stack([x_data] * num_points, axis=0)
+    epsilon_array = np.linspace(min_epsilon, max_epsilon,
+                                num_points)
+
+    # Get the height, width and number of channels
+    shape = x.get_shape().as_list()
+    height = shape[1]
+    width = shape[2]
+    channels = shape[3]
+    size = height * width * channels
+
+    epsilon_array_t = tf.constant(epsilon_array, dtype=tf.float32)
+    epsilon_array_t = tf.stack([epsilon_array_t] * size, axis=1)
+    epsilon_array_t = tf.reshape(epsilon_array_t,
+                                 shape=[num_points, height,
+                                        width, channels])
+
+    adv_x_epsilon = x + eta * epsilon_array_t
+    logits = model.get_logits(adv_x_epsilon)
+    with sess.as_default():
+        log_prob_adv_array = sess.run(logits, feed_dict={x: x_data})
+    return log_prob_adv_array
+
+
+def linear_extrapolation_plot(log_prob_adv_array, y, file_name,
+                              min_epsilon=-10, max_epsilon=10,
+                              num_points=21):
+    """Generate linear extrapolation plot.
+
+    Args:
+        log_prob_adv_array: Numpy array containing log probabilities
+        y: Tf placeholder for the labels
+        file_name: Plot filename
+        min_epsilon: Minimum value of epsilon over the interval
+        max_epsilon: Maximum value of epsilon over the interval
+        num_points: Number of points used to interpolate
+    """
+    import matplotlib.pyplot as plt
+
+    figure = plt.figure()
+    figure.canvas.set_window_title('Cleverhans: Linear Extrapolation Plot')
+
+    correct_idx = np.argmax(y, axis=0)
+    fig = plt.figure()
+    plt.xlabel('Epsilon')
+    plt.ylabel('Log probabilities')
```

Comment:
They're actually logits, not log probabilities, right?

Reply:
Fixed